lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb

```diff
@@ -146,12 +180,14 @@ def disallows_and_double_checks_value_of!(value, message)
         end
 
         def disallows_original_or_typecast_value?(value, message)
-          disallows_value_of(blank_value, @expected_message)
+          disallows_value_of(value, message)
         end
 
         def blank_value
           if collection?
             []
+          elsif expects_to_allow_nil?
+            ''
```

Comment:
Maybe need a better implementation?

Reply:
Now that I think about it, perhaps the matcher should work like this?

* If the matcher is qualified with `allow_nil`, then it should test that empty-string and nil are disallowed.
* However, if it _is_ qualified with `allow_nil`, then it should only test that empty-string is disallowed (because it tests that nil is allowed).

What do you think about that?

Reply:
IMO, both a blank string and a `nil` value return `true` to `#blank?`,
so as long as the `ActiveModel::Validations::PresenceValidator` uses `#blank?` to check the validity of an attribute, both a true blank string and a `nil` can be considered a blank string (when the attribute's type is either `String` or `Nil`).

Reply:
> Perhaps what you mean is, when the matcher is not qualified with allow_nil it should disallow both `''` and `nil`?

Yes, that's what I meant :)

> IMO, both a blank string and a nil value return true to #blank?

Right, exactly. As it stands, the presence matcher only uses `nil` in making its assertion, and in your changes here you conditionally flip that to an empty string, but really, in the default case where `allow_nil` is not specified, the presence matcher needs to make use of both, since PresenceValidator is not just testing that the attribute is nil but is testing that it is blank.

Reply:
What do you think about this:

```diff
 def matches?(subject)
   super(subject)

   if secure_password_being_validated?
     ignore_interference_by_writer.default_to(when: :blank?)
     disallows_and_double_checks_value_of!(blank_value, @expected_message)
-  elsif expects_to_allow_nil?
-    allows_value_of(nil) &&
-      disallows_original_or_typecast_value?(blank_value, @expected_message)
   else
-    disallows_original_or_typecast_value?(blank_value, @expected_message)
+    disallowed_values
+      .map { |value| disallows_original_or_typecast_value?(value, @expected_message) }
+      .reduce { |prev, curr| prev && curr }
   end
 end
```

```diff
-def blank_value
+def disallowed_values
+  disallowed = []
+
   if collection?
-    []
+    disallowed << []
   elsif expects_to_allow_nil?
-    ''
+    disallowed << ''
   else
-    nil
+    disallowed << nil
+    disallowed << ''
   end
+
+  disallowed
 end
```

Reply:
That's good! Perhaps instead of `reduce` you could use `all?` And I don't think you have to write any tests to cover this -- we are changing how the matcher works internally, and as long as these changes still make the current tests pass then you should be good.

Reply:
I have tried to use the code similar to the previous example to see if it still passes the tests.
However, after I ran the specs, there is one failure. And I think the cause might be testing agains an empty string when not qualified with `allow_nil`.
So I checked out `c0960bd7` (the commit right before this PR), and made the following modification only:
```diff
 def blank_value
   if collection?
     []
   else
-    nil
+     ''
   end
 end
```
After I ran the specs again, there are several failed specs expecting an assertion to fail with a message but the assertion did not fail. 
So I think maybe we should not explicitly disallow an empty string when not using `allow_nil`.

Reply:
Also, there are several places that have the following example:
```ruby
it_supports(
  'ignoring_interference_by_writer',
  tests: {
    accept_if_qualified_but_changing_value_does_not_interfere: {
      changing_values_with: :nil_to_blank
    },
    reject_if_qualified_but_changing_value_interferes: {
      model_name: 'Example',
      attribute_name: :attr,
      changing_values_with: :never_falsy,
      expected_message: <a message>
    }
  }
)
```
Though I don't fully understand how this works, I think this might already have covered the case where setting an attribute to an empty string: `changing_values_with: :nil_to_blank `.

Reply:
Can you go ahead and push up your modifications so that I can see the test failures? That way I can debug them for you.

Also in response to your last comment, that test is making sure that if you have an attribute that changes nil to an empty string when it is assigned, that this won't cause the presence validation to fail. I guess it's true that this is implicitly testing the empty string case, but I think it would be more clear to have an explicit test case / context so that it would be easier to find if we have to do so later.

Reply:
Sure, I've just pushed my code.